Feature email verify

buchholz-sean opened the pull request:
User interaction for `!verify` command. Includes basic code generation that can be swapped out for something simpler or something more cryptographically secure if needed (see @MROwens Code Generation PR). `TODO`s indicate where further enhancements will be added.

buchholz-sean committed:
Create basic verif code generation

Created `generateCode` method to create random *n*-digit verification
code. Discord messageCollector checks for user response that matches
generated code.

Not implemented yet:

- Client-side email validation
- Method to email code to user
- Method to add verified user to db
- Method to ping Officer or higher on verify failure

buchholz-sean committed:
Add user feedback to collector

Provides feedback to the user on collecting the verification code or on
timing out without receiving the code.

Needs adjustment to run collector in DM channel instead of in text
channel to avoid user confusion.

buchholz-sean committed:
Implement client-side email validation

Ensure user's email address domain matches one of the approved patterns
with basic conditional logic. If not a valid email address, notify user.

MROwens commented on bot/controllers/verify.js:
```diff
@@ -0,0 +1,99 @@
+          if (n > max) {
+            return generateCode(max) + generateCode(n - max);
+          }
+          max = 10 ** (n + add);
```
When I ran this on my machine I got an unexpected token error for the double `**` in `max = 10 ** (n + add)`. 
Would suggest switching it to `max = 10 * (n + add)`.

buchholz-sean replied:
The `**` operator is exponents, so `10 ** (n + add)` is 10<sup>(n + add)</sup>. `10 * (n + add)` isn't equivalent and I don't know how it would affect the randomness of the resultant string. I'm only using this operator because the linter rejected `Math.pow()`. 

Can you give more detail on the error you got? Was it from the linter or something else?

Ultimately though if we wind up using a different code generator method this becomes a non-issue.

MROwens replied:
Sure thing. This is what I got. 
'0|max      | /Users/Matt_Owens/Desktop/DWA/Max-Bot/bot/controllers/verify.js:30
0|max      |           max = 10 ** (n + add);
0|max      |                             ^
0|max      | SyntaxError: Unexpected token *
0|max      |     at Object.exports.runInThisContext (vm.js:53:16)
0|max      |     at Module._compile (module.js:513:28)
0|max      |     at Object.Module._extensions..js (module.js:550:10)
0|max      |     at Module.load (module.js:458:32)
0|max      |     at tryModuleLoad (module.js:417:12)
0|max      |     at Function.Module._load (module.js:409:3)
0|max      |     at Module.require (module.js:468:17)
0|max      |     at require (internal/module.js:20:19)
0|max      |     at fs.readdirSync.forEach (/Users/Matt_Owens/Desktop/DWA/Max-Bot/bot/controllers/index.js:22:18)
0|max      |     at Array.forEach (native)'

MROwens replied:
Sure thing. This is what I got. 
```javascript
0|max      | /Users/Matt_Owens/Desktop/DWA/Max-Bot/bot/controllers/verify.js:30
0|max      |           max = 10 ** (n + add);
0|max      |                     ^
0|max      | SyntaxError: Unexpected token *
0|max      |     at Object.exports.runInThisContext (vm.js:53:16)
0|max      |     at Module._compile (module.js:513:28)
0|max      |     at Object.Module._extensions..js (module.js:550:10)
0|max      |     at Module.load (module.js:458:32)
0|max      |     at tryModuleLoad (module.js:417:12)
0|max      |     at Function.Module._load (module.js:409:3)
0|max      |     at Module.require (module.js:468:17)
0|max      |     at require (internal/module.js:20:19)
0|max      |     at fs.readdirSync.forEach (/Users/Matt_Owens/Desktop/DWA/Max-Bot/bot/controllers/index.js:22:18)
0|max      |     at Array.forEach (native)
````

Rhuansantos replied:
@buchholz-sean your logic seems right!

ES5 `Math.pow(10, n + add)`
ES6 `10 ** (n + add)`

when I run both I have the same result however for some reason the ES6 version is erroring out!

MROwens commented:
Everything else looks great!

buchholz-sean committed:
Change !verify response type

Switched `resType` from `dm` to `reply` and flagged `allowInDm` as
`false` to ensure `messageCollector` is only operating on a server
channel. Minor syntax changes to bot responses to reflect message
context (i.e., avoid awkward repetition of user's name).

Rhuansantos reviewed (changes requested):
@buchholz-sean For now I will be hard-coding that part and merging this branch into dev so I can start working on how we will store that data!

Rhuansantos committed:
temporary fixing hard-coding string generator values

ghost commented:
## Code Gen
I've locally added the randomstring package via npm (is working well), other code was clashing in a few environments.

## !verify via Direct Message
Would suggest looking into allowing the ```!verify``` command to be the only thing entered that then fires off a PM/DM from Max-Bot for sensitive information such as email address.
This will also ensure server channels don't get spammed (win win).

## feedback for users
Would suggest looking into giving feedback to user if the code entered was wrong and only allowing 3 attempts before flagging an officer for manual approval vs. waiting for time to expire.

joegreen2 commented on bot/controllers/verify.js:
```diff
@@ -0,0 +1,115 @@
+          const number = Math.floor(Math.random() * (max - (min + 1))) + min;
+          return ('' + number).substring(add);
+        };
+        // We can set `codeLength` to whatever length we want the verif code to be.
```
small typo "verify"

joegreen2 commented on bot/controllers/verify.js:
```diff
@@ -0,0 +1,115 @@
+          collector.on('collect', (m) => {
+            util.log('Collected', m.content, 3);
+            // TODO: Database integration/email verif process
+            message.reply('Thanks! I\'ll get to work adding you the servers right away!');
```
small typo '... adding you "to" the servers...'
